### PR TITLE
#41 api authentication - coding style

### PR DESCRIPTION
**Pull request by Jokeswar**

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -22,7 +22,7 @@ Vagrant.configure("2") do |config|
   config.vm.provider :vmck do |vmck|
     vmck.image_path = 'cluster-master.qcow2.tar.gz'
     vmck.vmck_url = ENV['VMCK_URL']
-    vmck.memory = 1024 * 2
-    vmck.cpus = 1
+    vmck.memory = 1024 * 4
+    vmck.cpus = 2
```
Comment (mgax) [inline Vagrantfile]: Do we need to double the resource use? Why?

File: vmck/api.py [hunk 2]
```diff
@@ -23,6 +28,26 @@ def home(request):
     })
 
 
+def create_token(request):
+    auth_header = request.headers['Authorization']
+    (token_type, _, credentials) = auth_header.partition(' ')
+
+    (username, password) = base64.b64decode(credentials).decode('latin-1').split(':')  # noqa: E501
+    # TODO if password is not correct return failure
```
Comment (mgax) [inline vmck/api.py]: We need to authenticate the user. Like in `acs-interface`.

File: vmck/api.py [hunk 3]
```diff
@@ -23,6 +28,26 @@ def home(request):
     })
 
 
+def create_token(request):
+    auth_header = request.headers['Authorization']
+    (token_type, _, credentials) = auth_header.partition(' ')
```
Comment (mgax) [inline vmck/api.py]: This looks weird. What does this bit of code do?

File: vmck/middleware.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import jwt
+
+from django.http import JsonResponse
+
+from vmck import settings
+
+
+def jwt_authentication_middleware(get_response):
+    def middleware(request):
+        if request.path in settings.UNAUTHENTICATED_PATHS or settings.DEBUG:
```
Comment (mgax) [inline vmck/middleware.py]: I don't think `DEBUG = True` should mean "skip authorization", is it really necessary?

File: vmck/middleware.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            return get_response(request)
+
+        try:
+            auth_header = request.headers['Authorization']
+            if not auth_header.startswith('Bearer'):
```
Comment (mgax) [inline vmck/middleware.py]: This looks like a good candidate for regular expressions. It's way more readable and safer than splitting strings.

File: testsuite/test_api.py [hunk 6]
```diff
@@ -1,7 +1,10 @@
-from time import time, sleep
+import base64
 import subprocess
+from time import time, sleep
+
 import pytest
-from django.conf import settings # noqa
+from django.conf import settings  # noqa
```
Comment (mgax) [inline testsuite/test_api.py]: Why are we importing settings and then saying `noqa`? It seems pointless to import settings.

File: testsuite/test_api.py [hunk 7]
```diff
@@ -19,15 +22,26 @@ def __init__(self, client):
         self.client = client
 
     def start(self):
-        resp = self.client.post('/v0/jobs', content_type='application/json')
+        credentials = base64.b64encode(b'test:test').decode('latin1')
```
Comment (mgax) [inline testsuite/test_api.py]: Please extract this encode-decode dance into a helper function.

File: testsuite/test_api.py [hunk 8]
```diff
@@ -19,15 +22,26 @@ def __init__(self, client):
...
+        auth_resp = self.client.get('/v0/token',
+                                    HTTP_AUTHORIZATION=f'Basic {credentials}').json()  # noqa: E501
```
Comment (mgax) [inline testsuite/test_api.py]: Ugh, why not indent this nicely?

```python
auth_resp = self.client.get(
  '/v0/token',
  HTTP_AUTHORIZATION=f'Basic {credentials}',
).json()
```

File: testsuite/test_api.py [hunk 9]
```diff
@@ -36,7 +50,10 @@ def wait(self, timeout=900):
             sleep(1)
 
     def destroy(self):
-        return self.client.delete(self.url)
+        return self.client.delete(
+            self.url,
+            HTTP_AUTHORIZATION=f'Bearer {self.token}',
```
Comment (mgax) [inline testsuite/test_api.py]: It should not be necessary to pass `HTTP_AUTHORIZATION` all the time, there must be a way to set default headers on the client object. A quick look at [the source code](https://docs.djangoproject.com/en/2.2/_modules/django/test/client/#Client) suggests there's a `client.defaults` dictionary for that.

File: testsuite/settings.py [hunk 10]
```diff
@@ -17,6 +19,6 @@
 NOMAD_JOB_PREFIX = f'testsuite-{random_code(8)}-'
 NOMAD_DEPLOYMENT_NAME = f"test-vmck@{hostname} {datetime.now().strftime('%H:%M:%S')}"  # noqa: E501
 
-VMCK_BACKEND = os.environ.get('TESTING_BACKEND', 'docker')
+VMCK_BACKEND = os.environ.get('TESTING_BACKEND', 'qemu')
```
Comment (mgax) [inline testsuite/settings.py]: Why are we changing this?

File: testsuite/settings.py [hunk 11]
```diff
@@ -17,6 +19,6 @@
...
 
-QEMU_IMAGE_PATH_PREFIX = os.environ.get('TESTING_QEMU_IMAGE_PATH_PREFIX')
+QEMU_IMAGE_PATH_PREFIX = os.environ.get('TESTING_QEMU_IMAGE_PATH_PREFIX', 'http://10.42.1.1:10001')  # noqa: E501
```
Comment (mgax) [inline testsuite/settings.py]: Why are we changing this?

File: pytest.ini [hunk 12]
```diff
@@ -1,3 +1,4 @@
 [pytest]
 DJANGO_SETTINGS_MODULE = testsuite.settings
+log_cli = true
```
Comment (mgax) [inline pytest.ini]: I'm not a fan of this setting. pytest already captures logs and prints them for failed tests. Let's find another way to fix your problem.

File: ci/deploy-vmck [hunk 13]
```diff
@@ -7,4 +7,3 @@ trap "vagrant destroy -f"  EXIT
 vagrant up
 
 vagrant ssh -- /opt/vmck/ci/test-pytest.sh
-vagrant ssh -- /opt/vmck/ci/test-alive-on-cluster.sh
```
Comment (mgax) [inline ci/deploy-vmck]: Why is this line going away?

**Comment by Jokeswar** [comment]
Not up to date.

**Pull request closed without merging**